compiler/rustc_const_eval/src/interpret/call.rs

```diff
@@ -86,6 +86,9 @@ impl<'tcx, M: Machine<'tcx>> InterpCx<'tcx, M> {
                 let (_, field) = layout.non_1zst_field(self).unwrap();
                 self.unfold_transparent(field, may_unfold)
             }
+            ty::Pat(base, _) => self.layout_of(*base).expect(
```

Comment:
This should only be changed here if we also change the docs to guarantee that a pattern type is ABI-compatible with its inner type.

Reply:
Well, considering it's the only way to guarantee that `NonZeroU8` is still compatible with `u8`, I think the answer is yes, and that we already have at least tests for that. I'll add additional tests for bare pattern types

Reply:
Please make a note on the tracking issue so that before stabilization, we decide whether this is true for all pattern types or only for some.

Reply:
We unconditionally allow transmuting from pattern types to their base type. For any pattern. I don't see a design that could work differently

Reply:
"A can be transmuted to B" does *not* imply "A and B are ABI-compatible". For instance, `i32` and `u32` are not ABI-compatible. The code you changed here is about ABI compatibility.